### DISCUSSION TITLE
Issue with OGR Array in Postgres Foreign Table using ogr_fdw

### DISCUSSION THREAD

**Discussion by LuGIS92** [pramsey/pgsql-ogr-fdw#254] [Q&A] [answered]
Hello everyone,

Firstly, I want to express my gratitude for the fantastic work on the ogr_fdw extension. It's an invaluable tool for integrating geospatial data into my Postgres databases.

I'm encountering an issue while connecting to a WFS server using ogr_fdw. The aim is to access planning data from a municipality, but I'm running into a snag with a specific column.

Here's how I set up the server and foreign table:
`CREATE SERVER biedenkopf_wfs
    FOREIGN DATA WRAPPER ogr_fdw
    OPTIONS (datasource 'WFS:https://geoservices.marburg-biedenkopf.de/wfs/bauleitplanung/06534004?', format 'WFS', config_options 'CPL_DEBUG=ON ARRAY_AS_STRING=YES');

CREATE FOREIGN TABLE IF NOT EXISTS hessen.bplan_biedenkopf(
    fid bigint,
    geometry geometry(MultiSurface,3857),
    gml_id character varying COLLATE pg_catalog."default",
    ags character varying COLLATE pg_catalog."default",
    akt_datenb character varying COLLATE pg_catalog."default",
    akt_datens character varying COLLATE pg_catalog."default",
    akt_recht character varying COLLATE pg_catalog."default",
    aktenz_k character varying COLLATE pg_catalog."default",
    aufstellb character varying COLLATE pg_catalog."default",
    bemerkung character varying COLLATE pg_catalog."default",
    ebene_plan character varying COLLATE pg_catalog."default",
    einleitb character varying COLLATE pg_catalog."default",
    feststellb character varying COLLATE pg_catalog."default",
    flaechengr character varying COLLATE pg_catalog."default",
    gem_nr character varying COLLATE pg_catalog."default",
    gemarkung character varying COLLATE pg_catalog."default",
    gemeinde character varying COLLATE pg_catalog."default",
    hintergr character varying COLLATE pg_catalog."default",
    hintergr_a character varying COLLATE pg_catalog."default",
    insp_id character varying COLLATE pg_catalog."default",
    kreis character varying COLLATE pg_catalog."default",
    land character varying COLLATE pg_catalog."default",
    leb_geo character varying COLLATE pg_catalog."default",
    medien character varying COLLATE pg_catalog."default",
    name character varying COLLATE pg_catalog."default",
    name_a character varying COLLATE pg_catalog."default",
    nutzung character varying COLLATE pg_catalog."default",
    offenlegb character varying COLLATE pg_catalog."default",
    offenlegz character varying COLLATE pg_catalog."default",
    plan_ae character varying COLLATE pg_catalog."default",
    plan_af_tp character varying COLLATE pg_catalog."default",
    plan_e character varying COLLATE pg_catalog."default",
    plan_id character varying COLLATE pg_catalog."default",
    rechts_ver character varying COLLATE pg_catalog."default",
    rechtskr character varying COLLATE pg_catalog."default",
    regbez character varying COLLATE pg_catalog."default",
    regnum character varying COLLATE pg_catalog."default",
    satzbeschl character varying COLLATE pg_catalog."default",
    status character varying COLLATE pg_catalog."default",
    typ character varying COLLATE pg_catalog."default",
    uplan character varying COLLATE pg_catalog."default",
    utm_nord character varying COLLATE pg_catalog."default",
    utm_ost character varying COLLATE pg_catalog."default",
    pdf character varying[] COLLATE pg_catalog."default"
)
    SERVER biedenkopf_wfs
    OPTIONS (layer 'mabi:bauleitplanung');
`

When I attempt to select data from this table, I get the following error:

`ERROR: column "pdf" of foreign table "bplan_biedenkopf" uses an OGR array, currently unsupported`

The "pdf" column is supposed to store links to downloadable PDF files and seems to be causing the issue. I've tried several configurations, including setting ARRAY_AS_STRING=YES, but the error persists. Excluding this column does fix the error, but the links in the "pdf" column are crucial for my project.When I load the WFS in QGIS for testing purposes, the issue with the column does not occur.

I would greatly appreciate any advice on handling or circumventing this OGR array issue. If anyone could suggest a workaround or solution to include the "pdf" column, it would be incredibly helpful.

Thank you so much for your time and help!

Lukas

**Comment by pramsey** [comment, marked as answer]
Well, ogr_fdw definitely does not support arrays, so there's no configuration fix at that level. Your attempt with `ARRAY_AS_STRING` is along the correct path, try to get OGR to paper over the difference. I'd tend to think that a server definition using that option combined with a table definition that just says the column type is 'varchar' would be the thing most likely to work. I'd experiment with ogr2ogr or ogrinfo a little to make sure you're actually getting plain strings out of OGR, as getting exactly the right option string into exactly the right OGR variable can be tricky.

**Reply by pramsey** [reply]
(Array support should not be hard to add, it's just that I never saw an OGR source with an array typed column before, so I didn't want to spend the effort.)

**Reply by LuGIS92** [reply]
Dear Paul,

I wanted to share some great news - the issue I was facing with the OGR array in the ogr_fdw extension has been successfully resolved! Following your advice, I experimented a bit and found that defining the 'pdf' column as text in the foreign table worked wonders (in combination with ARRAY_AS_STRING).
I appreciate your guidance immensely. Thanks again for your help and for contributing to such a useful tool!

**Comment by pramsey** [comment]
See #98

**Reply by LuGIS92** [reply]
Thanks Paul for implementing it, I will test it as soon as possible!
